components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/builders/X509CredentialImpl.java

```diff
@@ -43,8 +60,183 @@
+        } catch (Exception e) {
+            throw new IdentityException("Error retrieving private key and the certificate for tenant " +
+                    MultitenantConstants.SUPER_TENANT_DOMAIN_NAME, e);
+        }
+    }
+
+    /**
+     * Set private key and X509Certificate from the Sign KeyStore which is defined under Security.SAMLSignKeyStore
+     * in carbon.xml.
+     *
+     * @throws IdentityException
+     */
+    private void initializeCredentialForSuperTenantFromSignKeyStore() throws IdentityException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Initializing Key Data for super tenant using separate sign key store.");
+        }
+
+        try {
+            if (superTenantSignKeyStore == null) {
+                String keyStoreLocation = ServerConfiguration.getInstance().getFirstProperty(
+                        SECURITY_SAML_SIGN_KEY_STORE_LOCATION);
+                try (FileInputStream is = new FileInputStream(keyStoreLocation)) {
+                    String keyStoreType = ServerConfiguration.getInstance().getFirstProperty(
+                            SECURITY_SAML_SIGN_KEY_STORE_TYPE);
+                    KeyStore keyStore = KeyStore.getInstance(keyStoreType);
+                    char[] keyStorePassword = ServerConfiguration.getInstance().getFirstProperty(
+                            SECURITY_SAML_SIGN_KEY_STORE_PASSWORD).toCharArray();
+                    keyStore.load(is, keyStorePassword);
+                    superTenantSignKeyStore = keyStore;
+                } catch (IOException e) {
+                    throw new IdentityException("Unable to read keystore.", e);
+                } catch (CertificateException e) {
+                    throw new IdentityException("Unable to read certificate.", e);
+                }
+            }
+
+            String keyAlias = ServerConfiguration.getInstance().getFirstProperty(
+                    SECURITY_SAML_SIGN_KEY_STORE_KEY_ALIAS);
+            char[] keyPassword = ServerConfiguration.getInstance().getFirstProperty(
+                    SECURITY_SAML_SIGN_KEY_STORE_KEY_PASSWORD).toCharArray();
+            Key privateKey = superTenantSignKeyStore.getKey(keyAlias, keyPassword);
+            Certificate publicKey = superTenantSignKeyStore.getCertificate(keyAlias);
+
+            if (privateKey instanceof PrivateKey) {
+                this.privateKey = (PrivateKey) privateKey;
+            } else {
+                throw new IdentityException("Configured signing KeyStore private key is invalid.");
+            }
+
+            if (publicKey instanceof X509Certificate) {
+                this.signingCert = (X509Certificate) publicKey;
+            } else {
+                throw new IdentityException("Configured signing KeyStore X509Certificate is invalid.");
+            }
+        } catch (NoSuchAlgorithmException e) {
+            throw new IdentityException("Unable to load algorithm", e);
```

Comment:
Let's add more information. May be algorithm name, key alias, or anything related

Reply:
Also this method looks too long. If we can split this into multiple readable methods, that would be great !